coderd/database/queries/workspaces.sql

```diff
@@ -161,6 +178,30 @@ WHERE
 			name ILIKE '%' || @name || '%'
 		ELSE true
 	END
+	-- Filter by agent status
+	-- has-agent: is only applicable for workspaces in "start" transition. Stopped and deleted workspaces don't have agents.
+	AND CASE
+		WHEN @has_agent :: text != '' THEN
+			latest_build.transition = 'start'::workspace_transition
+			AND CASE
+				WHEN @has_agent = 'timeout' THEN
+					latest_build.first_connected_at IS NULL AND (latest_build.created_at + latest_build.connection_timeout_seconds * INTERVAL '1 second' < NOW())
+				WHEN @has_agent = 'connecting' THEN
+			    	latest_build.first_connected_at IS NULL
+				WHEN @has_agent = 'disconnected' THEN
+					(
+						latest_build.disconnected_at IS NOT NULL AND
+						latest_build.disconnected_at > latest_build.last_connected_at
+					) OR (
+						latest_build.last_connected_at IS NOT NULL AND
+						latest_build.last_connected_at + 6 * INTERVAL '1 second' < NOW() -- FIXME agentInactiveDisconnectTimeout = 6
```

Comment:
I think you could just do `INTERVAL '6 seconds'` instead. Also, is this continually updated?

Reply:
The interval should be passed from [agentInactiveDisconnectTimeout](https://github.com/coder/coder/blob/2b6c229e4e2e94776500fe366b538afb6c24c199/coderd/coderd.go#L78) for every `Workspaces()`, at least for ones using `HasAgent`.

I set it temporarily to [6 seconds](https://github.com/coder/coder/blob/2b6c229e4e2e94776500fe366b538afb6c24c199/coderd/coderd.go#L118-L123) to show the overall PR idea. I can improve it now, but I'm not convinced as it introduces a strict dependency on Coder configuration.

Reply:
All FIXMEs are implemented now.